src/charmed_kubeflow_chisme/testing/_unit_tests.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""
+# Testing
+
+Tools for unit or integration testing, such as importable and reusable tests.
+
+## Usage example
+
+```python
+import pytest
+from charm import Operator
+from charmed_kubeflow_chisme.testing import test_leadership_events as leadership_events
+from charmed_kubeflow_chisme.testing import test_missing_relation as missing_relation
+from ops.model import WaitingStatus
+from ops.testing import Harness
+
+
+@pytest.fixture
+def harness():
+    return Harness(Operator)
+
+
+def test_leadership_events(harness):
+    leadership_events(harness)
+
+
+def test_missing_relation(harness):
+    missing_relation(harness, WaitingStatus, oci_image_added=False)
+```
+"""
+
+from contextlib import nullcontext as does_not_raise
+
+import pytest
+from oci_image import MissingResourceError
+from ops.model import BlockedStatus, WaitingStatus
+
+
+def test_leadership_events(harness):
+    """Test leader-elected event handling.
+
+    Tests if the unit raises correct status:
+
+    * `WaitingStatus` when it's not a leader
+
+    * when elected as the leader, it should leave `WaitingStatus`
+
+    * `WaitingStatus` when another unit is elected as the leader.
+
+    Args:
+        harness: instantiated Charmed Operator Framework test harness
+    """
+    harness.set_leader(False)
+    harness.begin_with_initial_hooks()
+    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+    harness.set_leader(True)
+    assert harness.charm.model.unit.status != WaitingStatus("Waiting for leadership")
+    harness.set_leader(False)
+    # Emit another leader_elected event due to https://github.com/canonical/operator/issues/812
+    harness._charm.on.leader_elected.emit()
+    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+
+
+def test_missing_image(harness, expected_status=BlockedStatus, leader_check=True):
+    """Tests if the unit raises an expected status when a required oci image is missing
+    in a charm with the following checks order:
+
+    1) check for leadership (optional)
+
+    2) check oci image
+
+    Args:
+        harness: instantiated Charmed Operator Framework test harness
+        expected_status: a subclass of `ops.model.StatusBase`. Default: `BlockedStatus`
+        leader_check: whether the unit should be set to leader first. Default: True
+    """
+    if leader_check:
+        harness.set_leader(True)
+    harness.begin_with_initial_hooks()
```

Comment:
Thinking about this test case a bit more, we might not need it at all. The check you are expecting to run here is `_check_image_details()`, I believe, which ensures there are valid image details for setting up the pod spec of certain charms. 

Some observations:
1) We are moving away from pod spec, meaning we eventually won't need to check for image details at all
2) This test case is only calling `set_leader` and `begin_with_initial_hooks()`, and then asserting a status
    2.1) It seems we are only wrapping those calls in a test case
    2.2) This test case is not checking whether the image details are valid (which is the goal of the check you are depending on here)
3) This test case depends in the order of checks in the charm, thus if someone decides to change that order, this test case will potentially fail or throw false-positives.

That being said, I suggest you reconsider even having it.